### ISSUE TITLE
Types are not checked when building `case` expression

### ISSUE THREAD

**Issue by HaoYang670** [apache/arrow-datafusion#4660]
**Describe the bug**
The type of `case` expression is not checked when building the plan.

**To Reproduce**
```sql
DataFusion CLI v15.0.0
❯ create table t as select 2 as a;
0 rows in set. Query took 0.013 seconds.
❯ select case
    when a = 1 then true
    when a = 2 then 'abc'
    else 123
  end  
from t;
Internal("The type of Boolean AND Utf8 of binary physical should be same")
```

**Expected behavior**
We should return a `DatafusionError::Plan(...)`, instead of an `Internal` error.

**Additional context**
The `CaseBuilder` does some type checking, but it seems that it is bypassed.
https://github.com/apache/arrow-datafusion/blob/c2f199adbec8b650de5668c97675f8a757607df8/datafusion/expr/src/conditional_expressions.rs#L100-L110

- [ ] Check `when` expr
- [ ] Check `then` and `else`

**Comment by HaoYang670** [comment]
Blocked by #4615

**Comment by HaoYang670** [comment]
test on pg: https://onecompiler.com/postgresql/3z4rdp5g9

**Comment by HaoYang670** [comment]
Close it as this has been fixed by #5820.
